subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationRoles.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.configurations;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public enum ConfigurationRoles implements ConfigurationRole {
+    /**
+     * An unrestricted configuration, which can be used for any purpose.
+     *
+     * This is available for backwards compatibility, but should not be used for new configurations.  It is
+     * the default role for configurations created when another more specific role is <strong>not</strong> specified.
+     */
+    @Deprecated
+    LEGACY(true, true, true, true, true, true),
+
+    /**
+     * Meant to be used only for consumption by other projects.
+     */
+    INTENDED_CONSUMABLE(true, false, false, false, false, false),
+
+    /**
+     * Meant to be used only for resolving dependencies.
+     */
+    INTENDED_RESOLVABLE(false, true, false, false, false, false),
```

Comment:
Hi Cédric, thanks for the early feedback.  This is still very much a work in progress, but at the moment my thoughts are that the three `INTENDED_` roles seem to be of limited use for replacing existing configurations.  Most existing configurations seem to be serving multiple purposes.  I expect in most places we'll have to begin by substituting one of the 2 `DEPRECATED_` roles instead.

Temporarily adding a `INTENDED_RESOLVABLE_WITH_DEPENDENCIES` which allows declaring against and resolving without emitting deprecation warnings isn't out of the question.  There is still much to be done first to get this working and evaluate the impact.

Reply:
@melix - You might be interested in seeing the new spec for this, it's linked in the description and should be publicly readable.

Reply:
I'm not sure we really considered the use cases for a consumable/bucket as much as we did for the resolvable/bucket. Though, we definitely see the value in only needing to make a single configuration for the resolvable/bucket use-case, as needing to create two would be cumbersome much of the time -- especially in buildscripts. 

@melix Could you potentially clarify some of the use cases where you think a consumable/bucket would be useful for users?

Reply:
Added `INTENDED_RESOLVABLE_BUCKET` role.

Reply:
> @melix Could you potentially clarify some of the use cases where you think a consumable/bucket would be useful for users?

Note that there is always the possibility to extend from another configuration, we're discussing whether it's convenient or not to be able to directly add dependencies to a consumable configuration. I am pretty sure I have done this before but can't recall an exact use case.

So an example of that would be when a project needs to expose dependencies to consumers, but not necessarily artifacts. For example it would be a safe way to "expose the dependencies" that a project uses. We're actually using something similar in the Micronaut internal build plugins, in order to be able to build an aggregated javadoc. We have a configuration which exposes the dependencies of each project. Although in this particular case, the configuration is declared to extend from the runtime dependencies, so it's not a bucket per se.

So basically I think this can be useful if you want to expose, in some way, a list of dependencies to a consumer, and that this list is _not_ directly a list available via another configuration. There shouldn't be many use cases for this.

Reply:
I think this is a good example of why we left open the possibility for defining custom roles for now, which should be able to handle infrequent use cases like this.  When we start moving this into the public API, I agree it would be convenient if there was some way to quickly create pairs of linked configurations for the bucket + resolvable and perhaps bucket + consumable cases.

I think is worth more thought at the time these methods begin to migrate to the public API, but this can remain as-is for now for internal usage.  Thanks for sharing the use case.